File: MasterProject/Code/biotrait_clustering.py
import pandas as pd
import numpy as np
from openai import OpenAI
import hdbscan
from sklearn.preprocessing import normalize


client = OpenAI(api_key="")
DATA_PATH = "/Users/danielzhu/Documents/Master Project/Data/GlobalDataset.csv"
RESULTS_PATH = "/Users/danielzhu/Documents/Master Project/Results"

df = pd.read_csv(DATA_PATH, encoding="ISO-8859-1", low_memory=False)
df = df.drop_duplicates(subset=["OriginalTraitName", "OriginalTraitDef"])
df = df.fillna("")
df = df.reset_index(drop=True)

print(f"Unique (name+def) pairs: {len(df)}")


def build_text(row, version):
    parts = [f"Trait: {row['OriginalTraitName']}."]
    if version >= 2 and row['OriginalTraitDef']:
        parts.append(f"Definition: {row['OriginalTraitDef']}.")
    if version >= 3 and row['OriginalTraitUnit']:
        parts.append(f"Unit: {row['OriginalTraitUnit']}.")
    if version >= 4 and row['ConOrder']:
        parts.append(f"Organism order: {row['ConOrder']}.")
    if version >= 5 and row['ConFamily']:
        parts.append(f"Organism family: {row['ConFamily']}.")
    if version >= 6 and row['ConStage']:
        parts.append(f"Life stage: {row['ConStage']}.")
    return " ".join(parts)

# ── 3. Embedding ─────────────────────────────────────────────────
def get_embeddings(texts, model="text-embedding-3-large", batch_size=100):
    all_embeddings = []
    for i in range(0, len(texts), batch_size):
        batch = texts[i:i+batch_size]
        print(f"  Batch {i//batch_size + 1}/{(len(texts)-1)//batch_size + 1}...")
        response = client.embeddings.create(input=batch, model=model)
        all_embeddings.extend([item.embedding for item in response.data])
    return np.array(all_embeddings)

# ── 4. loop ─────────────────────────────────────────────────────
results = []

for v in range(1, 7):
    print(f"\n=== Version {v} ===")
    
    texts = [build_text(row, v) for _, row in df.iterrows()]
    embeddings = get_embeddings(texts)
    embeddings_norm = normalize(embeddings)
    
    np.save(f"{RESULTS_PATH}/biotraits_embeddings_v{v}.npy", embeddings_norm)
    
    clusterer = hdbscan.HDBSCAN(
        min_cluster_size=5,
        min_samples=3,
        metric="euclidean",
        gen_min_span_tree=True
    )
    cluster_labels = clusterer.fit_predict(embeddings_norm)
    
    n_clusters  = len(set(cluster_labels)) - (1 if -1 in cluster_labels else 0)
    n_noise     = int(sum(cluster_labels == -1))
    noise_rate  = n_noise / len(cluster_labels)
    
    sizes = pd.Series(cluster_labels)
    sizes = sizes[sizes != -1].value_counts()
    avg_size = float(sizes.mean()) if len(sizes) > 0 else 0
    
    print(f"  Clusters: {n_clusters}, Noise: {n_noise} ({noise_rate:.1%})")
    
    results.append({
        "version":    f"v{v}",
        "x":          v,
        "n_clusters": n_clusters,
        "n_noise":    n_noise,
        "noise_rate": noise_rate,
        "avg_size":   avg_size
    })
    
    df_out = df.copy()
    df_out["text"] = texts
    df_out["version"] = f"v{v}"
    df_out["cluster_id"] = cluster_labels
    df_out.to_csv(f"{RESULTS_PATH}/biotraits_clustering_v{v}.csv", index=False)


results_df = pd.DataFrame(results)
results_df.to_csv(f"{RESULTS_PATH}/biotraits_rarefaction.csv", index=False)

print("\n=== Summary ===")
print(results_df.to_string(index=False))

File: MasterProject/Code/Biotrait_clustering.R
library(reticulate)
use_condaenv("pymc-env", required = TRUE)
py_config()
# ============================================================
# BioTraits GlobalDataset — Clustering Pipeline (V1-V6)
# ============================================================
library(tidyverse)
library(reticulate)


# Sys.setenv(RETICULATE_PYTHON = "/path/to/your/python")  

setwd("/Users/danielzhu/Documents/Master Project")


biotraits <- read_csv(
  "Data/GlobalDataset.csv",
  locale = locale(encoding = "ISO-8859-1"),
  show_col_types = FALSE
)


unique_pairs <- biotraits %>%
  mutate(across(c(OriginalTraitName, OriginalTraitDef, OriginalTraitUnit,
                  ConOrder, ConFamily, ConStage),
                ~str_trim(as.character(.)))) %>%
  distinct(OriginalTraitName, OriginalTraitDef, .keep_all = TRUE) %>%
  select(OriginalTraitName, OriginalTraitDef, OriginalTraitUnit,
         ConOrder, ConFamily, ConStage)

cat(sprintf("Unique (name+def) pairs: %d\n", nrow(unique_pairs)))


build_text <- function(df, version) {
  txt <- paste0("Trait: ", df$OriginalTraitName, ".")
  
  if (version >= 2) {
    def <- ifelse(is.na(df$OriginalTraitDef) | df$OriginalTraitDef == "",
                  "", paste0(" Definition: ", df$OriginalTraitDef, "."))
    txt <- paste0(txt, def)
  }
  if (version >= 3) {
    unit <- ifelse(is.na(df$OriginalTraitUnit) | df$OriginalTraitUnit == "",
                   "", paste0(" Unit: ", df$OriginalTraitUnit, "."))
    txt <- paste0(txt, unit)
  }
  if (version >= 4) {
    ord <- ifelse(is.na(df$ConOrder) | df$ConOrder == "",
                  "", paste0(" Organism order: ", df$ConOrder, "."))
    txt <- paste0(txt, ord)
  }
  if (version >= 5) {
    fam <- ifelse(is.na(df$ConFamily) | df$ConFamily == "",
                  "", paste0(" Organism family: ", df$ConFamily, "."))
    txt <- paste0(txt, fam)
  }
  if (version >= 6) {
    stg <- ifelse(is.na(df$ConStage) | df$ConStage == "",
                  "", paste0(" Life stage: ", df$ConStage, "."))
    txt <- paste0(txt, stg)
  }
  txt
}


openai <- import("openai")
hdbscan <- import("hdbscan")
np <- import("numpy")
sklearn_prep <- import("sklearn.preprocessing")

client <- openai$OpenAI(api_key = "")

get_embeddings <- function(texts, model = "text-embedding-3-large", batch_size = 100) {
  all_emb <- list()
  n <- length(texts)
  for (i in seq(1, n, by = batch_size)) {
    idx <- i:min(i + batch_size - 1, n)
    cat(sprintf("  Batch %d/%d...\n", ceiling(i/batch_size), ceiling(n/batch_size)))
    resp <- client$embeddings$create(input = as.list(texts[idx]), model = model)
    batch_emb <- lapply(resp$data, function(x) x$embedding)
    all_emb <- c(all_emb, batch_emb)
  }
  do.call(rbind, lapply(all_emb, unlist))
}


results <- list()

for (v in 1:6) {
  cat(sprintf("\n=== Version %d ===\n", v))
  
  texts <- build_text(unique_pairs, v)
  
  # embedding
  emb <- get_embeddings(texts)
  emb_norm <- sklearn_prep$normalize(emb)
  

  np$save(sprintf("Results/biotraits_embeddings_v%d.npy", v), emb_norm)
  
  # HDBSCAN
  clusterer <- hdbscan$HDBSCAN(
    min_cluster_size = 5L,
    min_samples = 3L,
    metric = "euclidean",
    gen_min_span_tree = TRUE
  )
  labels <- clusterer$fit_predict(emb_norm)
  labels <- as.integer(labels)
  

  n_clusters <- length(unique(labels[labels != -1]))
  n_noise    <- sum(labels == -1)
  noise_rate <- n_noise / length(labels)
  
  sizes <- as.data.frame(table(labels[labels != -1]))
  avg_size <- mean(sizes$Freq)
  
  cat(sprintf("  Clusters: %d, Noise: %d (%.1f%%)\n",
              n_clusters, n_noise, noise_rate * 100))
  
  results[[v]] <- tibble(
    version    = sprintf("v%d", v),
    x          = v,
    n_clusters = n_clusters,
    n_noise    = n_noise,
    noise_rate = noise_rate,
    avg_size   = avg_size
  )
  

  out <- unique_pairs %>%
    mutate(text = texts, version = sprintf("v%d", v), cluster_id = labels)
  write_csv(out, sprintf("Results/biotraits_clustering_v%d.csv", v))
}


summary_df <- bind_rows(results)
write_csv(summary_df, "Results/biotraits_rarefaction.csv")

cat("\n=== Summary ===\n")
print(summary_df)
